eth/stagedsync/stage_execute.go

```diff
@@ -422,3 +460,88 @@ func min(a, b uint64) uint64 {
 	}
 	return b
 }
+
+func NewTouchReaderWriter(r state.StateReader, w state.WriterWithChangeSets) *TouchReaderWriter {
+	return &TouchReaderWriter{
+		r:                r,
+		w:                w,
+		readCodes:        make(map[common.Address]struct{}),
+		updatedCodes:     make(map[common.Address]struct{}),
+		createdContracts: make(map[common.Address]struct{}),
+	}
+}
+
+type TouchReaderWriter struct {
+	r                state.StateReader
+	w                state.WriterWithChangeSets
+	readCodes        map[common.Address]struct{}
+	updatedCodes     map[common.Address]struct{}
+	createdContracts map[common.Address]struct{}
+}
+
+func (d *TouchReaderWriter) ReadAccountData(address common.Address) (*accounts.Account, error) {
+	return d.r.ReadAccountData(address)
+}
+
+func (d *TouchReaderWriter) ReadAccountStorage(address common.Address, incarnation uint64, key *common.Hash) ([]byte, error) {
+	return d.r.ReadAccountStorage(address, incarnation, key)
+}
+
+func (d *TouchReaderWriter) ReadAccountCode(address common.Address, incarnation uint64, codeHash common.Hash) ([]byte, error) {
+	d.readCodes[address] = struct{}{}
+	return d.r.ReadAccountCode(address, incarnation, codeHash)
+}
+
+func (d *TouchReaderWriter) ReadAccountCodeSize(address common.Address, incarnation uint64, codeHash common.Hash) (int, error) {
+	return d.r.ReadAccountCodeSize(address, incarnation, codeHash)
+}
+
+func (d *TouchReaderWriter) ReadAccountIncarnation(address common.Address) (uint64, error) {
+	return d.r.ReadAccountIncarnation(address)
+}
+
+func (d *TouchReaderWriter) WriteChangeSets() error {
+	return d.w.WriteChangeSets()
+}
+
+func (d *TouchReaderWriter) WriteHistory() error {
+	return d.w.WriteHistory()
+}
+
+func (d *TouchReaderWriter) UpdateAccountData(ctx context.Context, address common.Address, original, account *accounts.Account) error {
+	return d.w.UpdateAccountData(ctx, address, original, account)
+}
+
+func (d *TouchReaderWriter) UpdateAccountCode(address common.Address, incarnation uint64, codeHash common.Hash, code []byte) error {
+	d.updatedCodes[address] = struct{}{}
+	return d.w.UpdateAccountCode(address, incarnation, codeHash, code)
+}
+
+func (d *TouchReaderWriter) DeleteAccount(ctx context.Context, address common.Address, original *accounts.Account) error {
+	return d.w.DeleteAccount(ctx, address, original)
+}
+
+func (d *TouchReaderWriter) WriteAccountStorage(ctx context.Context, address common.Address, incarnation uint64, key *common.Hash, original, value *uint256.Int) error {
+	return d.w.WriteAccountStorage(ctx, address, incarnation, key, original, value)
+}
+
+func (d *TouchReaderWriter) CreateContract(address common.Address) error {
+	d.createdContracts[address] = struct{}{}
+	return d.w.CreateContract(address)
+}
+
+func (d *TouchReaderWriter) AllCodes() map[common.Address]struct{} {
+	c := make(map[common.Address]struct{}, len(d.readCodes)+len(d.updatedCodes)+len(d.createdContracts))
+
+	for addr := range d.readCodes {
```

Comment:
Why would we need to register read codes?

Reply:
Just implemented for all possible cases. At the moment only create contract is used.